File: sorting/LC121_best_time_to_buy_and_sell_stock/solution.cpp
#include <iostream>

class Solution {
public:
  int maxProfit(std::vector<int>& prices) {
    int lowest = INT_MAX;
    int profit = 0;
    int potentialProfit = 0;
    
    for(int i = 0; i < prices.size(); i++){
      if(prices[i] < lowest){
        lowest = prices[i];
      }
      potentialProfit = prices[i] - lowest;
      if(profit < potentialProfit){
        profit = potentialProfit;
      }
    }
    return profit;
  }
};

int main() {
  std::vector<int> prices = {7,1,5,3,6,4};
  Solution s;
  int output = s.maxProfit(prices);
  std::cout << output << std::endl;
}

File: sorting/LC121_best_time_to_buy_and_sell_stock/solution.ts
function maxProfit(prices: number[]): number {
  let profit: number = 0;
  let buyPrice: number = prices[0];
  for (let price of prices) {
      if (price > buyPrice) {
          profit = Math.max(profit, price - buyPrice);
      } else {
          buyPrice = price;
      }
  }
  return profit;
}

const main = () => {
  const prices: number[] = [7, 1, 5, 3, 6, 4];
  console.log(maxProfit(prices));
}

